proposals/exception-handling/Exceptions-formal-overview.md

```diff
@@ -0,0 +1,211 @@
+
+Throw Contexts
+
+```
+T ::= val* [_] instr* | label_n{instr*} T end | caught_m{a val^n} T end
+   | frame_n{F} T end
+```
+
+### Instructions
+
+
+```
+F; throw x  ↪  F; throw a  (if F.module.exnaddrs[x]=a)
+
+caught_m{a val^n} B^l[rethrow l] end
+  ↪ caught_m{a val^n} B^l[val^n (throw a)] end
+
+caught_m{a val^n} val^m end  ↪  val^m
+```
+
+A missing tagaddr in a `catch_m` clause (i.e., `a? = ε`) represents a `catch_all`.
+
+```
+F; val^n (try bt instr* (catch x instr'*)* (catch_all instr''*)? end)
+  ↪  F; label_m{} (catch_m{a instr'*}*{instr''*}? val^n instr* end) end
+  (if bt = [t1^n]→[t2^m] ∧ (val:t1)^n ∧ (F.module.exnaddrs[x]=a)*)
+
+catch_m{a? instr*}* val^m end ↪ val^m
+
+S; F; catch_m{a1? instr*}{a'? instr'*}* T[val^n (throw a)] end
+  ↪  S; F; caught_m{a val^n} (val^n)? instr* end
+  (if (a1? = ε ∨ a1? = a) ∧ S.tags(a).type = [t^n]→[])
+
+S; F; catch_m{a1? instr*}{a'? instr'*}* T[val^n (throw a)] end
+  ↪  S; F; catch_m{a'? instr'*}* T[val^n (throw a)] end
+  (if a1? ≠ ε ∧ a1? ≠ a)
+
+S; F; catch_m T[val^n (throw a)] end
+  ↪  S; F; val^n (throw a)
+
+
+val^n (try bt instr* delegate l)
+  ↪ label_m{} (delegate{l} val^n instr* end) end
+  (if bt = [t^n]→[t^m])
+
+delegate{l} val^n end ↪ val^n
+
+B^l[ delegate{l} T[val^n (throw a)] end ]
+  ↪ val^n (throw a)
+```
+
+### Typing rules for administrative instructions
+
+```
+S.tags[a].type = [t2*]→[]
+--------------------------------
+S;C ⊢ throw a : [t1* t2*]→[t*]
+
+(S.tags[a].type = [t'*]→[] ∧
+ S;C, labels {result [t*], kind catch} ⊢ instr1* : [t'*]→[t*])*
+(S;C, labels {result [t*], kind catch} ⊢ instr2* : []→[t*])?
+S;C, labels {result [t*], kind try} ⊢ instr3* : []→[t*]
+-----------------------------------------------------------------------
+S;C, labels [t*] ⊢ catch_n{a instr1*}*{instr2*}? instr3* end : []→[t*]
+
+S;C, labels {result [t*], kind try} ⊢ instr* : []→[t*]
+C.labels[l].kind = try
+-----------------------------------------------------------------------
+S;C, labels [t*] ⊢ delegate{l} instr* end : []→[t*]
+
+S.tags[a].type = [t'*]→[]
+(val:t')*
+S;C, labels {result [t*], kind catch} ⊢ instr* : []→[t*]
+--------------------------------------------------------------------------------
+S;C, labels {result [t*], kind catch} ⊢ caught_m{a val^n} instr* end : []→[t'*]
+```
+
+By adding the attribute `kind` to labels, we are creating situations in the proof of type preservation, where we have a derivation of some `S;C, label [t*] ⊢ instr* : []→[t*]` but we need to have that `S;C, label {result [t*], kind <labelkind>} ⊢ instr* : []→[t*]` for some `<labelkind> ::= try | catch`. To resolve this we add the following typing rule for labels in the context, which ensures our newly introduced `try` and `catch` blocks can contain any instructions a regular block can.
+
+```
+S;C, label [t*] ⊢ instr* : []→[t*]
+labelkind = try ∨ labelkind = catch
+-----------------------------------------------------------
+S;C, label {result [t*], kind labelkind} ⊢ instr* : []→[t*]
+```
```

Comment:
Hm, can you give an example of why this rule is needed? AFAICS, labelkinds try or catch are strictly more permissive than a regular label, so why would you ever need to forget them?